[SPARK-49558][SQL] Add SQL pipe syntax for LIMIT/OFFSET and ORDER/SORT/CLUSTER/DISTRIBUTE BY

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR adds SQL pipe syntax support for LIMIT/OFFSET and ORDER/SORT/CLUSTER/DISTRIBUTE BY.

For example:

```
CREATE TABLE t(x INT, y STRING) USING CSV;
INSERT INTO t VALUES (0, 'abc'), (1, 'def');

TABLE t
|> ORDER BY x
|> LIMIT 1 OFFSET 1

1	def
```

### Why are the changes needed?

The SQL pipe operator syntax will let users compose queries in a more flexible fashion.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds a few unit test cases, but mostly relies on golden file test coverage. I did this to make sure the answers are correct as this feature is implemented and also so we can look at the analyzer output plans to ensure they look right as well.

### Was this patch authored or co-authored using generative AI tooling?

No

dtenedor committed:
commit

dtenedor commented:
cc @cloud-fan @gengliangwang this is the PR to support LIMIT/OFFSET + sorting. There are a few more changes in the `AstBuilder` for this one but still contained only in the parser.

dtenedor committed:
fix test

cloud-fan commented on sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql:
```diff
@@ -571,6 +583,95 @@ table t
+-- Multiple order by clauses are not supported in the same pipe operator
+table t
+|> order by x desc order by x + y
+order by y;
```
Do we need this final order by to trigger the error?

cloud-fan commented on sql/core/src/test/resources/sql-tests/results/pipe-operators.sql.out:
```diff
@@ -1673,6 +1691,279 @@ org.apache.spark.sql.catalyst.ExtendedAnalysisException
+table t
+|> where x = 1
+|> select y
+|> offset 1
```
this doesn't fail?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -1010,21 +1018,40 @@ class AstBuilder extends DataTypeAstBuilder
+    }
+    if (forPipeOperators && windowClause != null) {
+      throw QueryParsingErrors.combinationQueryResultClausesUnsupportedError(
+        ctx, s"WINDOW clauses within SQL pipe operators")
```
This doesn't really match `COMBINATION_QUERY_RESULT_CLAUSES`, shall we add a new error condition for it?

dtenedor replied:
Sounds good, done.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -1010,21 +1018,40 @@ class AstBuilder extends DataTypeAstBuilder
     val withOffset = withWindow.optional(offset) {
+      if (forPipeOperators && clause.nonEmpty) {
+        throw QueryParsingErrors.combinationQueryResultClausesUnsupportedError(
+          ctx, s"the $clause and $offsetClause clauses")
```
shall we highlight it's only a limitation in the SQL pipe syntax?

dtenedor replied:
Sounds good, done.

cloud-fan commented on sql/core/src/test/resources/sql-tests/results/pipe-operators.sql.out:
```diff
@@ -1673,6 +1691,279 @@ org.apache.spark.sql.catalyst.ExtendedAnalysisException
+  "errorClass" : "UNSUPPORTED_FEATURE.COMBINATION_QUERY_RESULT_CLAUSES",
+  "sqlState" : "0A000",
+  "messageParameters" : {
+    "clauses" : "ORDER BY/SORT BY/DISTRIBUTE BY/CLUSTER BY"
```
To match this existing error message, shall we change the pipe one to be `ORDER BY/LIMIT within pipe operator`?

dtenedor replied:
Sounds good, done.

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

dtenedor reviewed (commented):
Thanks @cloud-fan for your review!

dtenedor commented:
The most recent CI failure is not related

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -1010,21 +1018,44 @@ class AstBuilder extends DataTypeAstBuilder
-      throw QueryParsingErrors.combinationQueryResultClausesUnsupportedError(ctx)
+      val allClauses = "ORDER BY/SORT BY/DISTRIBUTE BY/CLUSTER BY"
+      if (forPipeOperators) {
+        throw QueryParsingErrors.clausesWithPipeOperatorsUnsupportedError(ctx, allClauses)
```
We should always call `combinationQueryResultClausesUnsupportedError` here, as this is not related to pipe. We don't support this combination in both pipe and classic SQL.

dtenedor replied:
Sounds good, this is done.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -1010,21 +1018,44 @@ class AstBuilder extends DataTypeAstBuilder
+      val limitClause = "LIMIT"
+      if (forPipeOperators && clause.nonEmpty && clause != offsetClause) {
+        throw QueryParsingErrors.clausesWithPipeOperatorsUnsupportedError(
+          ctx, s"the $clause and $limitClause clauses")
```
```suggestion
          ctx, s"the co-existence of $clause and $limitClause clauses")
```

dtenedor replied:
Sounds good, this is done.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -1010,21 +1018,44 @@ class AstBuilder extends DataTypeAstBuilder
     val withOffset = withWindow.optional(offset) {
+      if (forPipeOperators && clause.nonEmpty) {
+        throw QueryParsingErrors.clausesWithPipeOperatorsUnsupportedError(
+          ctx, s"the $clause and $offsetClause clauses")
```
```suggestion
          ctx, s"the co-existence of $clause and $offsetClause clauses")
```

dtenedor replied:
Sounds good, this is done.

cloud-fan commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -4926,9 +4926,14 @@
+      },
       "COMBINATION_QUERY_RESULT_CLAUSES" : {
         "message" : [
-          "Combination of ORDER BY/SORT BY/DISTRIBUTE BY/CLUSTER BY."
```
looking at the current code change, seems we don't need to update it anymore?

dtenedor replied:
Sounds good, this is done.

cloud-fan commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -4926,9 +4926,14 @@
           "Catalog <catalogName> does not support <operation>."
         ]
       },
+      "CLAUSE_WITH_PIPE_OPERATORS" : {
```
Maybe @srielau can also comment here, but it looks to me that we have two new errors:
1. WINDOW in the SQL pipe operator. We don't support window definition in SQL pipe yet.
2. More than one QUERY_RESULT_CLAUSES in SQL pipe. SQL pipe is designed to specify one operator at a time, so we don't allow any combination, such as ORDER BY col LIMIT 1, or LIMIT 1 OFFSET 1

srielau replied:
1. appears to be a 0A000 (unsupported feature)
2. Isn't that a syntax error?

cloud-fan replied:
yes, 2 is a syntax error for the pipe statement.

dtenedor replied:
Thanks, split this into two new errors per suggestion.

dtenedor reviewed (commented):
Thanks @cloud-fan for your reviews!

dtenedor committed:
respond to code review comments

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -1014,17 +1022,35 @@ class AstBuilder extends DataTypeAstBuilder
 
     // OFFSET
     // - OFFSET 0 is the same as omitting the OFFSET clause
+    val offsetClause = "OFFSET"
```
nit: shall we put all the constant clauses into an object?

dtenedor replied:
Sure, it is good to keep magic strings out of the code.

dtenedor committed:
respond to code review comments

cloud-fan commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -5006,6 +5011,11 @@
           "Multiple bucket TRANSFORMs."
         ]
       },
+      "MULTIPLE_QUERY_RESULT_CLAUSES_WITH_PIPE_OPERATORS" : {
```
shall we turn it into a top-level error class? This is syntax error, not unsupported feature, and should have a different SQL state.

dtenedor replied:
Sounds good, done.

cloud-fan commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -5006,6 +5011,11 @@
       },
+      "MULTIPLE_QUERY_RESULT_CLAUSES_WITH_PIPE_OPERATORS" : {
+        "message" : [
+          "Syntax error: the SQL pipe operator syntax using |> does not support <clauses>. Please separate the multiple result clauses into separate pipe operators and then retry the query again."
```
how about
```
<clause1> and <clause2> can not coexist in SQL pipe operator (1>). Please separate ...
```

cloud-fan replied:
The idea is to make the error parameters simple, so that people can consume it programatically.

dtenedor replied:
Sure, switched to this format.

dtenedor committed:
respond to code review comments

respond to code review comments

dtenedor reviewed (commented):
Thanks @cloud-fan again for your reviews!

cloud-fan commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -3448,6 +3448,11 @@
+  "MULTIPLE_QUERY_RESULT_CLAUSES_WITH_PIPE_OPERATORS" : {
+    "message" : [
+      "<clause1> and <clause2> can not coexist in the same SQL pipe operator using '|>'. Please separate the multiple result clauses into separate pipe operators and then retry the query again."
+    ]
```
we missed the sqlState

dtenedor replied:
👍 I added the sqlState in here.

dtenedor committed:
respond to code review comments

dtenedor committed:
commit

dtenedor committed:
commit

dtenedor committed:
Merge branch 'master' into pipe-order-by
sync

cloud-fan commented:
The linter failure is unrelated, I'm merging it to master, thanks!

Angryrou commented on sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql:
```diff
@@ -571,6 +583,97 @@ table t
+
+-- The WINDOW clause is not supported yet.
+table windowTestData
+|> window w as (partition by cte order by val)
```
Hi Daniel @dtenedor , I noticed that this window clause in Spark differs from what’s described in the original paper and documentation. Could you share your thoughts on this?

The [documentation](https://github.com/google/zetasql/blob/master/docs/pipe-syntax.md#window-pipe-operato
) specifies that a window operator should always include a window function with an OVER clause. However, in Spark's syntax, the window operator only returns a window definition without requiring an OVER clause.

I think it makes sense to keep the existing window syntax (as shown in this example) since the Extend clause will cover the window operator’s functionality as described in the paper. However, I’d like to confirm the expected behavior of the window clause in Spark SQL before proceeding with a PR.

Thanks in advance!

cloud-fan replied:
In Spark we provide a default window frame so the OVER clause becomes optional. We shouldn't change it.
